### ISSUE TITLE
Grain services documentation

### ISSUE THREAD

**Issue by DixonDs** [dotnet/orleans#3122]
Unless I am missing it, grain services feature is not mentioned in documentation. If that's right, it would be really great, if someone who knows it very well could add a relevant documentation page.

**Comment by sergeybykov** [comment]
@DixonDs  You are correct - there is still no documentation for grain services still. I guess I need to write it, unless @jamescarter-le (who implemented the feature) already did that and is keeping it to himself. :-)

**Comment by yevhen** [comment]
Eagerly awaiting ... ))

**Comment by yevhen** [comment]
Any updates?

**Comment by martinothamar** [comment]
There's an extension method on `ISiloHostBuilder`: `AddGrainService<SomeGrainService>()`. 
Type constraint is: `where T : GrainService`.
It ends up calling this bit: https://github.com/dotnet/orleans/blob/2f3a5863677963f05ad301659509baefa24bb5d7/src/Orleans.Runtime/Services/GrainServicesSiloBuilderExtensions.cs#L48

Basically the silo fetches `IGrainService` types from the service provider when starting: https://github.com/dotnet/orleans/blob/2f3a5863677963f05ad301659509baefa24bb5d7/src/Orleans.Runtime/Silo/Silo.cs#L554

**Comment by nockawa** [comment]
I've tried to create a GrainService and things said here are not totally complete.
In order for this to work you have to register **both** the Service **and** its Client.
The code looks something like this
```csharp
    var builder = new SiloHostBuilder()
        .AddGrainService<DataService>()  // Register GrainService
        .ConfigureServices(s =>
         {
            // Register Client of GrainService
            s.AddSingleton<IDataServiceClient, DataServiceClient>(); 
        })
```

**Comment by AlexanderButs** [comment]
Hi, guys.

Thanks for such a great feature. Just small question to You - can GrainService read from streams?

Thanks.

**Comment by sergeybykov** [comment]
>Thanks for such a great feature. Just small question to You - can GrainService read from streams?

No, they cannot subscribe to streams because grain services aren't virtual - their addresses are tied to physical silos underneath.

**Comment by AlexanderButs** [comment]
Thanks for so fast response.

Maybe it's incorrect place to ask but anyway - is there a way to have a grain per silo and call all those grains as a broadcast? I need to update inmemory state on each silo inside whole cluster.

Thanks once again.

**Comment by Zonciu** [comment]
Is that GrainService similar with `[StatelessWorker]` Grain which has constant identity(like `0`)?
I'm using `[StatelessWorker] QueryGrain` with id `0`  for client to query data, can GrainService do this?

**Comment by sergeybykov** [comment]
In the narrow sense of having one instance per silo, they are similar. But they are different in several other aspects - invocation API, auto-activation, repartitioning, etc.

**Comment by sergeybykov** [comment]
Resolved via #4946.
